compiler/rustc_hir_typeck/src/_if.rs

```diff
@@ -0,0 +1,297 @@
+
+#[derive(Clone, Copy, Debug)]
+pub(crate) struct IfExprParts<'tcx> {
+    pub cond: &'tcx hir::Expr<'tcx>,
+    pub then: &'tcx hir::Expr<'tcx>,
+    pub else_branch: Option<&'tcx hir::Expr<'tcx>>,
+}
+
+#[derive(Clone, Copy, Debug)]
+struct IfExprWithParts<'tcx> {
+    expr: &'tcx hir::Expr<'tcx>,
+    parts: IfExprParts<'tcx>,
+}
+
+#[derive(Clone, Debug, Default)]
+struct IfChain<'tcx> {
+    guarded_branches: SmallVec<[IfGuardedBranch<'tcx>; 4]>,
+    tail: IfChainTail<'tcx>,
+    error: Option<ErrorGuaranteed>,
+}
+
+impl<'tcx> IfChain<'tcx> {
+    fn last_expr(&self) -> Option<&'tcx hir::Expr<'tcx>> {
+        if let IfChainTail::FinalElse(final_else) = &self.tail {
+            final_else.expr.into()
+        } else {
+            self.guarded_branches.last().map(|l| l.expr_with_parts.expr)
+        }
+    }
+}
+
+#[derive(Clone, Debug)]
+struct IfGuardedBranch<'tcx> {
+    expr_with_parts: IfExprWithParts<'tcx>,
+    cond_diverges: Diverges,
+    body: BranchBody<'tcx>,
+}
+
+#[derive(Clone, Debug)]
+enum IfChainTail<'tcx> {
+    FinalElse(BranchBody<'tcx>),
+    Missing,
+}
+
+#[derive(Clone, Debug)]
+struct BranchBody<'tcx> {
+    expr: &'tcx hir::Expr<'tcx>,
+    ty: Ty<'tcx>,
+    diverges: Diverges,
+    span: Span,
+}
+
+impl<'tcx> Default for IfChainTail<'tcx> {
+    fn default() -> Self {
+        IfChainTail::Missing
+    }
+}
+
+impl<'tcx> IfChainTail<'tcx> {
+    fn diverges(&self) -> Diverges {
+        match &self {
+            IfChainTail::FinalElse(else_branch) => else_branch.diverges,
+            IfChainTail::Missing => Diverges::Maybe,
+        }
+    }
+}
+
+impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
+    pub(crate) fn check_expr_if(
+        &self,
+        expr_id: HirId,
+        sp: Span,
+        parts: &IfExprParts<'tcx>,
+        orig_expected: Expectation<'tcx>,
+    ) -> Ty<'tcx> {
+        let root_if_expr = self.tcx.hir_expect_expr(expr_id);
+        let expected = orig_expected.try_structurally_resolve_and_adjust_for_branches(self, sp);
+
+        let initial_diverges = self.diverges.get();
+
+        let chain =
+            self.collect_if_chain(&IfExprWithParts { expr: root_if_expr, parts: *parts }, expected);
+
+        let coerce_to_ty = expected.coercion_target_type(self, sp);
+        let mut coerce: DynamicCoerceMany<'_> = CoerceMany::new(coerce_to_ty);
+
+        let tail_defines_return_position_impl_trait =
+            self.return_position_impl_trait_from_match_expectation(orig_expected);
+
+        for (idx, branch) in chain.guarded_branches.iter().enumerate() {
+            if idx > 0 {
+                let merged_ty = coerce.merged_ty();
+                self.ensure_if_branch_type(branch.expr_with_parts.expr.hir_id, merged_ty);
+            }
+
+            let branch_body = &branch.body;
+            let next_else_expr = chain
+                .guarded_branches
+                .get(idx + 1)
+                .map(|next| next.expr_with_parts.expr)
+                .or(chain.last_expr().into());
```

Comment:
Hmm, this doesn't seem right (or at least, is too subtle for my liking): when idx is len-1, `chain.last_expr()` is called and there is no final else, uses the *current* expr (since it is the last one).